pkg/proxy/ipvs/proxier.go

```diff
@@ -1972,7 +1972,15 @@ func (proxier *Proxier) syncEndpoint(svcPortName proxy.ServicePortName, onlyNode
 	} else {
 		clusterEndpoints, localEndpoints, _, _ := proxy.CategorizeEndpoints(endpoints, svcInfo, proxier.nodeLabels)
 		if onlyNodeLocalEndpoints {
-			endpoints = localEndpoints
+			if len(localEndpoints) > 0 {
+				endpoints = localEndpoints
+			} else {
+				// https://github.com/kubernetes/kubernetes/pull/97081
+				// Allow access from local PODs even if no local endpoints exist.
+				// Traffic from an external source will be routed but the reply
+				// will have the POD address and will be discarded.
```

Comment:
> but the reply will have the POD address and will be discarded.

I'm still scratching my head on this one to be honest, and I think I'm just missing something.

Why are replies to external sources discarded? Wouldn't conntrack keep track of the NAT entry here and know to map the reply packet from the Pod back to the external source? This is effectively how the VIPs work when externalTrafficPolicy=Cluster, so what is happening differently here that changes that behavior (maybe an iptables rule that that is specific to Local)?

Reply:
IIUC, this bypasses SNAT, so the eventual reply will go to the original client (DSR), not the proxying node.  That client THINKS it is talking to the service IP but the response comes from the pod IP.

Reply:
I see -- and this is safe since the fallback only happens when traffic policy is "Local" so we're guarenteed that SNAT is never enabled.